.github/workflows/full_stack_tests.yml

```diff
@@ -189,105 +189,136 @@ jobs:
+
+      # Desktop Test (Attempt 2)
       - name: Run Desktop Acceptance Test ${{ matrix.suite.name }} (Attempt 2 with screen recording)
-        if: ${{ failure() }}
+        id: desktop-test-attempt-2
+        if: ${{ steps.desktop-test-attempt-1.outcome == 'failure' }}
         run: >
           set -o pipefail;
           VIDEO_RECORDING_IS_ENABLED=1
           xvfb-run -a --server-args="-screen 0, 1920x1080x24"
           python -m scripts.run_acceptance_tests
           --skip-build --suite=${{ matrix.suite.name }} --prod_env --server_log_level=info
           | tee desktop_acceptance_test_retry_${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}.log
+
       - name: Upload desktop acceptance retry test logs
-        if: ${{ failure() }}
+        if: ${{ steps.desktop-test-attempt-2.outcome == 'failure' }}
         uses: actions/upload-artifact@v4
         with:
           name: desktop-acceptance-test-retry-logs-${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}
           path: desktop_acceptance_test_retry_${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}.log
-      - name: Uploading Desktop test recordings as Artifacts
-        if: ${{ failure() }}
+      
+      - name: Upload desktop test recordings as an artifact
+        # We are checking for failure of desktop-test-attempt-1 and not desktop-test-attempt-2 because
+        # because we want to upload video recording regardless of whether desktop-test-attempt-2 passes or fails.
+        if: ${{ steps.desktop-test-attempt-1.outcome == 'failure' }}
         uses: actions/upload-artifact@v4
         with:
           name: desktop-acceptance-test-video-${{steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME}}
           path: /home/runner/work/oppia/oppia_full_stack_test_video_recordings/acceptance
-      - name: Uploading generated test to angular module mapping as an artifact
-        if: ${{ failure() }}
+
+        # TODO(#22667): We are running this step twice, as snapshot gets deleted if 2nd attempt passes.
+      - name: Upload diff screenshots as an artifact
+        if: ${{ steps.desktop-test-attempt-2.outcome == 'failure' }}
         uses: actions/upload-artifact@v4
         with:
-          name: generated-test-to-angular-module-mapping-${{steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME}}
-          path: /home/runner/work/oppia/oppia/core/tests/test-modules-mappings/acceptance/${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}.txt
+          name: diff-snapshots-desktop-attempt-2-${{steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME}}
+          # Note: The path must be kept in sync with the one specified in
+          # core/tests/puppeteer-acceptance-tests/utilities/common/puppeteer-utils.ts
+          path: /home/runner/work/oppia/oppia/core/tests/puppeteer-acceptance-tests/diff-snapshots
+
+      # Mobile Test (Attempt 1)
```

Comment:
Is it better to do this as a {desktop, mobile} matrix to avoid the repetition? The logic ought to be the same.

Reply:
I checked it. We are only allowed maximum of 256 runs in a strategy. And currently, we have 135+ acceptance tests. Separating mobile and desktop will exceed total number of tests more than 256.

Reply:
Sorry, I meant number of jobs. Currently we are 138 acceptance tests = 138 jobs. If we use 2 jobs for each acceptance test (each for desktop and mobile), that would be 138 x 2 acceptance tests = 276 jobs > 256. However, I think we can achieve it once we remove old acceptance tests and reduce the number of tests we have to less than 128.

Reply:
Oh, I see. OK, please file an issue for this change and ensure it's clear that it can only be done when the number is < 128 (link to a reference citing the max number of runs).

Another alternative may be to split into external-focused and internal-focused acceptance tests. Though this still has the problem of duplication, we might need to declare a common action.